rateLimiter/src/main/java/com/flipkart/varadhi/qos/entity/ClientHistory.java

```diff
@@ -0,0 +1,70 @@
+package com.flipkart.varadhi.qos.entity;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.ObjectUtils;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+@Slf4j
+public class ClientHistory<T> {
+    Map<String, Deque<T>> clientHistoryMap; // clientId to history records
+    private final int slots;
+
+    public ClientHistory(int historySlots) {
+        this.clientHistoryMap = new ConcurrentHashMap<>();
+        this.slots = historySlots;
+    }
+
+    public T getRecentRecord(String clientId) {
```

Comment:
getRecentRecord & getRecentRecordForAll methods shouldn't exist.

This class is also generalizing wrong param. ClientHistory<T> looks wrong. Because the T is actually known. it is the trafficData. What is not known to this class is how to use that historical data to aggregate and predict overall throughput. So prediction algorithm is a pluggable.

Then, this class only needs one method to get the predicted load. so `TrafficData getPredictedLoad` ? `TrafficData getApproxLoad()`? 
can't think of other names :stuck_out_tongue:

Reply:
Made an interface that takes the client history and returns what to add for each client. Right now it only returns recent load info for each client which is inside the controller window